Support for crud operations such as derived delete queries

I find the following scenario:

I try to delete a series of items through a repository using the following repository definition:

`````
package com.microsip.cloud.command.broker.repositories;

import com.microsip.cloud.command.broker.model.AbstractItem;
import org.springframework.data.hazelcast.repository.HazelcastRepository;
import org.springframework.data.repository.NoRepositoryBean;

import java.io.Serializable;
import java.util.List;
import java.util.Optional;

@NoRepositoryBean
public interface ItemRepository<T extends AbstractItem, ID extends Serializable> extends HazelcastRepository<T, ID> {

    List<T> findBySubscriptionId(Long subscriptionId);

    Optional<T> findOneByIdAndSubscriptionId(Long originalId, Long subscriptionId);

    Boolean existsByIdAndSubscriptionId(Long originalId, Long subscriptionId);

    List<T> deleteBySubscriptionId(Long subscriptionId);

    List<T> removeBySubscriptionId(Long subscriptionId);

}
`````

By extending the definition of the repository by a more concrete definition for example:
`````
package com.microsip.cloud.command.broker.repositories;

import com.microsip.cloud.command.broker.model.GrupoDeLinea;
import org.springframework.stereotype.Repository;

@Repository
public interface GruposDeLineaRepository extends ItemRepository<GrupoDeLinea, String> {
}
`````

When creating a test to do item deletion, for example:
`````
@Test
    public void deleteElementsBySubscriptionId() {
        log.debug(testName.getMethodName());
        String subscriptionId = "123";
        List<GrupoDeLinea> grupoDeLineas = items.stream().filter(g -> subscriptionId.equals(String.valueOf(g.getSubscriptionId()))).collect(Collectors.toList());
        log.debug("grupoDeLineas is " + grupoDeLineas);
        List<GrupoDeLinea> expected = new ArrayList<GrupoDeLinea>();
        for (GrupoDeLinea gl : grupoDeLineas) {
            GrupoDeLinea g = new GrupoDeLinea();
            BeanUtils.copyProperties(gl, g);
            expected.add(g);
        }
        log.debug("expected is " + expected);
        List<GrupoDeLinea> deleted = gruposDeLineaRepository.removeBySubscriptionId(Long.valueOf(subscriptionId));
        log.debug("deleted is " + deleted);
        assertThat(deleted).isNotEmpty();
        List<GrupoDeLinea> actual = gruposDeLineaRepository.findBySubscriptionId(Long.valueOf(subscriptionId));
        log.debug("actual is " + actual);
        assertThat(actual).isNullOrEmpty();
        checkAfterCounts(-1 * (expected.size()));
    }
`````

Here are some logs:
`````
17:03:40.401 [main] DEBUG c.m.c.c.b.r.GruposDeLineaRepositoryTests - deleteElementsBySubscriptionId
17:03:40.402 [main] DEBUG c.m.c.c.b.r.GruposDeLineaRepositoryTests - grupoDeLineas is [GrupoDeLinea{systemId='6a2ab8d025a44d0dbb9a727c674a7ff5', name='brzxeujm', id=-7492532207654639100, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='d33e4e315fae460992b146ed526c25af', name='xdkrsitmup', id=-1236910821149465453, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='4e276f15f64046d6b81010ffe51b12b9', name='k', id=-6869027627953464114, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='7c2177914dd14592b982ea0ba67eaafe', name='bcyqxstl', id=3797886144915866000, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='c9adb69c6acc4867a79ee9274be8f9b9', name='emcexbudmq', id=236, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}]
17:03:40.403 [main] DEBUG c.m.c.c.b.r.GruposDeLineaRepositoryTests - expected is [GrupoDeLinea{systemId='6a2ab8d025a44d0dbb9a727c674a7ff5', name='brzxeujm', id=-7492532207654639100, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='d33e4e315fae460992b146ed526c25af', name='xdkrsitmup', id=-1236910821149465453, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='4e276f15f64046d6b81010ffe51b12b9', name='k', id=-6869027627953464114, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='7c2177914dd14592b982ea0ba67eaafe', name='bcyqxstl', id=3797886144915866000, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='c9adb69c6acc4867a79ee9274be8f9b9', name='emcexbudmq', id=236, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}]
17:03:40.419 [main] DEBUG c.m.c.c.b.r.GruposDeLineaRepositoryTests - deleted is [GrupoDeLinea{systemId='d33e4e315fae460992b146ed526c25af', name='xdkrsitmup', id=-1236910821149465453, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}]
17:03:40.427 [main] DEBUG c.m.c.c.b.r.GruposDeLineaRepositoryTests - actual is [GrupoDeLinea{systemId='7c2177914dd14592b982ea0ba67eaafe', name='bcyqxstl', id=3797886144915866000, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='6a2ab8d025a44d0dbb9a727c674a7ff5', name='brzxeujm', id=-7492532207654639100, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='4e276f15f64046d6b81010ffe51b12b9', name='k', id=-6869027627953464114, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}, GrupoDeLinea{systemId='c9adb69c6acc4867a79ee9274be8f9b9', name='emcexbudmq', id=236, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}]

java.lang.AssertionError: 
Expecting null or empty but was:<[GrupoDeLinea{systemId='7c2177914dd14592b982ea0ba67eaafe', name='bcyqxstl', id=3797886144915866000, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z},
    GrupoDeLinea{systemId='6a2ab8d025a44d0dbb9a727c674a7ff5', name='brzxeujm', id=-7492532207654639100, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z},
    GrupoDeLinea{systemId='4e276f15f64046d6b81010ffe51b12b9', name='k', id=-6869027627953464114, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z},
    GrupoDeLinea{systemId='c9adb69c6acc4867a79ee9274be8f9b9', name='emcexbudmq', id=236, subscriptionId=123, version=0, createdAt=2017-07-26T22:03:40.334Z, lastModified=2017-07-26T22:03:40.334Z}]>
`````

What I expect is that all elements that match the query are deleted, but only a single element is deleted.

It is a bug, or it should work like other spring-data implementations, for example:
[data-mongo http://docs.spring.io/spring-data/data-mongo/docs/current/reference/html/#mongodb.repositories.queries.delete](http://docs.spring.io/spring-data/data-mongo/docs/current/reference/html/#mongodb.repositories.queries.delete)
[spring-data-delete-by-is-supported https://stackoverflow.com/questions/23723025/spring-data-delete-by-is-supported](https://stackoverflow.com/questions/23723025/spring-data-delete-by-is-supported)
[Example 6 from https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.core-concepts](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.core-concepts)

[**Example 80. Delete…By Query**](http://docs.spring.io/spring-data/data-mongo/docs/current/reference/html/#mongodb.repositories.queries.delete)
`````
public interface PersonRepository extends MongoRepository<Person, String> {

  List <Person> deleteByLastname(String lastname);

  Long deletePersonByLastname(String lastname);
}
`````

[**Example 6. Derived Delete Query**](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.core-concepts)
`````
public interface UserRepository extends CrudRepository<User, Long> {

  Long deleteByLastname(String lastname);

  List<User> removeByLastname(String lastname);

}
`````
Even when you try to create a delete method that returns long instead of the list of items you get the following error:
`````
java.lang.ClassCastException: com.microsip.cloud.command.broker.model.GrupoDeLinea cannot be cast to java.lang.Long
`````
Even reviewing the tests in spring data hazelcast I find the following scenario, where they eliminate element by element.
[**QueryIT Repository fragment code test deleteBy**](https://github.com/hazelcast/spring-data-hazelcast/blob/master/src/test/java/org/springframework/data/hazelcast/repository/query/QueryIT.java#L133)
`````
	@Test
	public void deleteByLastname() {
		assertThat("Map fully populated before", this.personMap.size(), equalTo(Oscars.bestActors.length));
		
		Person person = this.personRepository.deleteByLastname("abcdefghijklmnopqrstuvwxyz");
		
		assertThat("Delete for unmatched name does nothing to map", this.personMap.size(), equalTo(Oscars.bestActors.length));
		assertThat("Delete for unmatched name does nothing to @Repository", this.personRepository.count(), equalTo(Long.valueOf(Oscars.bestActors.length)));
		assertThat("Delete for unmatched name returns null", person, nullValue());
		
		// Spencer Tracy, 1937 & 1938
		int count = 0;
		while ( (person = this.personRepository.deleteByLastname("Tracy")) != null ) {
			count++;
		}
		
		assertThat("Delete for matched name removes from map", this.personMap.size(), equalTo(Oscars.bestActors.length - 2));
		assertThat("Delete for matched name removes from @Repository", this.personRepository.count(), equalTo(Long.valueOf(Oscars.bestActors.length - 2)));
		assertThat("Delete for matched name returns correct count", count, equalTo(2));
		assertThat("1937 deleted", this.personMap.get("1937"), nullValue());
		assertThat("1938 deleted", this.personMap.get("1938"), nullValue());
	}
`````

Are there plans to incorporate these features?
@em-chips @neilstevenson I will appreciate your support.

--- a/src/main/java/org/springframework/data/hazelcast/repository/query/HazelcastPartTreeQuery.java
+++ b/src/main/java/org/springframework/data/hazelcast/repository/query/HazelcastPartTreeQuery.java
@@ -134,31 +134,32 @@ public class HazelcastPartTreeQuery
     }
 
     /**
-     * <p>
      * Execute a "delete" query, not really a query more of an operation.
-     * </P>
      * <p>
-     * <B>NOTE:</B> Delete is not a collection operation, the return value is a single object that
-     * was deleted.
-     * </P>
-     * <P>Although the <I>find</I> operation returns an iterator, the result set is either empty
-     * or has one domain object in it. If there are multiple possible matches it's random which
-     * one of the matches is deleted.
-     * </P>
      *
      * @param query       The query to run
      * @param queryMethod Used here to find the type of object to match the query
-     * @return Query The individual entry deleted
+     * @return Collection of deleted objects or the number of deleted objects
      */
     private Object executeDeleteQuery(final KeyValueQuery<?> query, final QueryMethod queryMethod) {
 
         Iterable<?> resultSet = this.keyValueOperations.find(query, queryMethod.getEntityInformation().getJavaType());
         Iterator<?> iterator = resultSet.iterator();
 
-        if (iterator.hasNext()) {
-            return this.keyValueOperations.delete(iterator.next());
+        List<Object> result = new ArrayList<>();
+        while (iterator.hasNext()) {
+            result.add(this.keyValueOperations.delete(iterator.next()));
+        }
+
+        if (queryMethod.isCollectionQuery()) {
+            return result;
+        } else if (long.class.equals(queryMethod.getReturnedObjectType()) || Long.class
+                .equals(queryMethod.getReturnedObjectType())) {
+            return result.size();
         } else {
-            return null;
+            throw new UnsupportedOperationException(String.format(
+                    "Illegal returned type: %s. The operation 'deleteBy' accepts only 'long' and 'Collection' as the returned "
+                            + "object type", queryMethod.getReturnedObjectType()));
         }
     }
 

--- a/src/test/java/org/springframework/data/hazelcast/repository/query/QueryIT.java
+++ b/src/test/java/org/springframework/data/hazelcast/repository/query/QueryIT.java
@@ -33,6 +33,7 @@ import test.utils.repository.standard.PersonRepository;
 
 import javax.annotation.Resource;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.TreeSet;
@@ -148,28 +149,52 @@ public class QueryIT
     }
 
     @Test
-    public void deleteByLastname() {
-        assertThat("Map fully populated before", this.personMap.size(), equalTo(Oscars.bestActors.length));
+    public void deleteBy() {
+        // given
+        // fully populated map
 
-        Person person = this.personRepository.deleteByLastname("abcdefghijklmnopqrstuvwxyz");
+        // when
+        long deletedPersonsSize = this.personRepository.deleteByLastname("Tracy");
 
-        assertThat("Delete for unmatched name does nothing to map", this.personMap.size(), equalTo(Oscars.bestActors.length));
-        assertThat("Delete for unmatched name does nothing to @Repository", this.personRepository.count(),
-                equalTo((long) Oscars.bestActors.length));
-        assertThat("Delete for unmatched name returns null", person, nullValue());
-
-        // Spencer Tracy, 1937 & 1938
-        int count = 0;
-        while (this.personRepository.deleteByLastname("Tracy") != null) {
-            count++;
-        }
-
+        // then
         assertThat("Delete for matched name removes from map", this.personMap.size(), equalTo(Oscars.bestActors.length - 2));
         assertThat("Delete for matched name removes from @Repository", this.personRepository.count(),
                 equalTo((long) (Oscars.bestActors.length - 2)));
-        assertThat("Delete for matched name returns correct count", count, equalTo(2));
+        assertThat("Delete for matched name returns correct count", deletedPersonsSize, equalTo(2L));
         assertThat("1937 deleted", this.personMap.get("1937"), nullValue());
         assertThat("1938 deleted", this.personMap.get("1938"), nullValue());
+    }
+
+    @Test
+    public void deleteByReturnCollection() {
+        // given
+        // fully populated map
+
+        // when
+        Collection<Person> deletedPersons = this.personRepository.deleteByFirstname("Spencer");
+
+        // then
+        assertThat("Delete for matched name removes from map", this.personMap.size(), equalTo(Oscars.bestActors.length - 2));
+        assertThat("Delete for matched name removes from @Repository", this.personRepository.count(),
+                equalTo((long) (Oscars.bestActors.length - 2)));
+        assertThat("Delete for matched name returns correct count", deletedPersons.size(), equalTo(2));
+        assertThat("1937 deleted", this.personMap.get("1937"), nullValue());
+        assertThat("1938 deleted", this.personMap.get("1938"), nullValue());
+    }
+
+    @Test
+    public void deleteByUnmatched() {
+        // given
+        // fully populated map
+
+        // when
+        long deletedPersonsSize = this.personRepository.deleteByLastname("abcdefghijklmnopqrstuvwxyz");
+
+        // then
+        assertThat("Delete for unmatched name does nothing to map", this.personMap.size(), equalTo(Oscars.bestActors.length));
+        assertThat("Delete for unmatched name does nothing to @Repository", this.personRepository.count(),
+                equalTo((long) Oscars.bestActors.length));
+        assertThat("Delete for unmatched name returns null", deletedPersonsSize, equalTo(0L));
     }
 
     // First by ascending == Min

--- a/src/test/java/test/utils/repository/standard/PersonRepository.java
+++ b/src/test/java/test/utils/repository/standard/PersonRepository.java
@@ -63,7 +63,9 @@ public interface PersonRepository
 
     public Long countDistinctLastnameByFirstname(String firstname);
 
-    public Person deleteByLastname(String firstname);
+    public long deleteByLastname(String firstname);
+
+    public List<Person> deleteByFirstname(String firstname);
 
     public Person findFirstIdByOrderById();
 

